File: backend/internal/core/domain/gems/gem.go
package gems

import (
	"time"
)

type GemColor int64

const (
	Purple GemColor = 0
	Pink   GemColor = 1
	Blue   GemColor = 2
	Black  GemColor = 3
	Yellow GemColor = 4
	Green  GemColor = 5
)

type Gem struct {
	ID         int64      `db:"id" goqu:"skipinsert"`
	Message    string     `db:"message"`
	Latitude   float64    `db:"latitude"`
	Longitude  float64    `db:"longitude"`
	CreatorId  int64      `db:"creator_id"`
	CreatedAt  time.Time  `db:"created_at" goqu:"skipinsert"`
	ReceiverId int64      `db:"receiver_id"`
	ReceivedAt *time.Time `db:"received_at"`
	Color      GemColor   `db:"color"`
	Attachment []byte     `db:"attachment"`
}

File: frontend/src/interfaces/gem.ts
import { Base64String, LatLng, User } from "@/interfaces";
import { Dayjs } from "dayjs";

export interface Gem {
  id: number;
  message: string;
  position: LatLng;
  createdAt: Dayjs;
  createdBy: User;
  receiver: User;
  receivedAt: Dayjs | null;
  color: GemColor;
  attachment: Base64String;
}

export type HexCode = string;

export enum GemColor {
  PURPLE = "#BC88B8",
  PINK = "#ffc8dd",
  BLUE = "#99CEE0",
  BLACK = "#8B8B8B",
  YELLOW = "#F0D072",
  GREEN = "#99BC88",
}

export interface DropGemFormState {
  id?: number;
  message?: string;
  receiverId?: number;
  color: GemColor;
  attachment?: File;
}

export interface GemLogsWithFriend {
  friend: User;
  gems: Gem[];
}

export interface GemLogs {
  gemLogsMap: Map<number, GemLogsWithFriend>;
}
